sockeye/training.py

```diff
@@ -146,657 +130,848 @@ def sym_gen(seq_lens):
-
-                # learning rate adjustment
-                if self.lr_scheduler is not None:
-                    if issubclass(type(self.lr_scheduler), lr_scheduler.AdaptiveLearningRateScheduler):
-                        lr_adjusted = self.lr_scheduler.new_evaluation_result(has_improved)
-                    else:
-                        lr_adjusted = False
-                    if lr_adjusted and not has_improved:
-                        if lr_decay_param_reset:
-                            logger.info("Loading parameters from last best checkpoint: %d", best_checkpoint)
-                            self._load_params(output_folder, best_checkpoint)
-                        if lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_INITIAL:
-                            logger.info("Resetting optimizer states to initial")
-                            self.load_optimizer_states(os.path.join(output_folder, C.OPT_STATES_INITIAL))
-                        elif lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_BEST:
-                            logger.info("Resetting optimizer states to last best checkpoint: %d", best_checkpoint)
-                            self.load_optimizer_states(os.path.join(output_folder, C.OPT_STATES_BEST))
+                    checkpoint_state = CheckpointState(checkpoint=self.state.checkpoint, metric_val=m_val)
+                    self.model.optimizer.pre_update_checkpoint(checkpoint_state)
 
-                if has_improved:
-                    best_params_path = os.path.join(output_folder, C.PARAMS_BEST_NAME)
-                    if os.path.lexists(best_params_path):
-                        os.remove(best_params_path)
-                    actual_best_params_fname = C.PARAMS_NAME % best_checkpoint
-                    os.symlink(actual_best_params_fname, best_params_path)
-                    if lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_BEST:
-                        best_opt_states_fname = os.path.join(output_folder, C.OPT_STATES_BEST)
-                        if os.path.exists(best_opt_states_fname):
-                            os.remove(best_opt_states_fname)
-                        self.save_optimizer_states(best_opt_states_fname)
-                    train_state.num_not_improved = 0
-                else:
-                    train_state.num_not_improved += 1
-                    logger.info("Model has not improved for %d checkpoints", train_state.num_not_improved)
+                # (5) adjust learning rates
+                self._adjust_learning_rate(has_improved, lr_decay_param_reset, lr_decay_opt_states_reset)
 
-                if max_num_not_improved >= 0 and train_state.num_not_improved >= max_num_not_improved:
+                # (6) save training state
+                self._save_training_state(train_iter)
+
+                # (7) determine stopping
+                if 0 <= max_num_not_improved <= self.state.num_not_improved:
```

Comment:
maybe it's just me, but I find this much harder to parse than two separate >= clauses.

Reply:
I was annoyed by the IDE warning. If you feel strongly about this I can revert, otherwise I would leave as is.

Reply:
sure, was just a comment, feel free to leave it like that.

Reply:
FWIW I also prefer separate clauses (probably because of C backgound).

Reply:
Out of curiosity, what was the warning?